File: lib/herb/token.rb
# frozen_string_literal: true
# typed: true

module Herb
  class Token
    attr_reader :value #: String
    attr_reader :range #: Range
    attr_reader :location #: Location
    attr_reader :type #: String

    #: (String, Range, Location, String) -> void
    def initialize(value, range, location, type)
      @value = value
      @range = range
      @location = location
      @type = type
    end

    #: () -> serialized_token
    def to_hash
      {
        value: value,
        range: range&.to_a,
        location: location&.to_hash,
        type: type,
      } #: Herb::serialized_token
    end

    #: (?untyped) -> String
    def to_json(state = nil)
      to_hash.to_json(state)
    end

    #: () -> String
    def tree_inspect
      %("#{value.force_encoding("utf-8")}" #{location.tree_inspect})
    end

    #: () -> String
    def value_inspect
      if type == "TOKEN_EOF"
        "<EOF>".inspect
      else
        value.inspect
      end
    end

    #: () -> String
    def inspect
      %(#<Herb::Token type="#{type}" value=#{value_inspect} range=#{range.tree_inspect} start=#{location.start.tree_inspect} end=#{location.end.tree_inspect}>)
    end
  end
end

File: javascript/packages/core/src/token.ts
import { Range, SerializedRange } from "./range.js"
import { Location, SerializedLocation } from "./location.js"

export type SerializedToken = {
  value: string
  range: SerializedRange
  location: SerializedLocation
  type: string
}

export class Token {
  readonly value: string
  readonly range: Range
  readonly location: Location
  readonly type: string

  static from(token: SerializedToken) {
    return new Token(
      token.value,
      Range.from(token.range),
      Location.from(token.location),
      token.type,
    )
  }

  constructor(value: string, range: Range, location: Location, type: string) {
    this.value = value
    this.range = range
    this.location = location
    this.type = type
  }

  toHash(): SerializedToken {
    return {
      value: this.value,
      range: this.range?.toArray(),
      location: this.location?.toHash(),
      type: this.type,
    }
  }

  toJSON(): SerializedToken {
    return this.toHash()
  }

  treeInspect(): string {
    return `"${this.value}" ${this.location.treeInspectWithLabel()}`
  }

  valueInspect(): string {
    return this.type === "TOKEN_EOF"
      ? JSON.stringify("<EOF>")
      : JSON.stringify(this.value)
  }

  inspect(): string {
    return `#<Herb::Token type="${this.type}" value=${this.valueInspect()} range=${this.range.treeInspect()} start=${this.location.start.treeInspect()} end=${this.location.end.treeInspect()}>`
  }

  toString(): string {
    return this.inspect()
  }
}
